src/v/cloud_topics/batcher/batcher.cc

```diff
@@ -168,77 +173,127 @@ batcher<Clock>::upload_object(object_id id, iobuf payload) {
 
-            // NOTE(Evgeny): the main workflow looks like this:
-            // - remove expired write requests
-            // - collect write requests which can be aggregated/uploaded as L0
-            //   object
-            // - create 'aggregator' and fill it with write requests (the
-            //   requests which are added to the aggregator shouldn't be removed
-            //   from _pending list)
-            // - the 'aggregator' is used to generate L0 object and upload it
-            // - the 'aggregator' is used to acknowledge (either success or
-            //   failure) all aggregated write requests
-            //
-            // The invariants here are:
-            // 1. expired write requests shouldn't be added to the 'aggregator'
-            // 2. if the request is added to the 'aggregator' its promise
-            //    shouldn't be set
-            //
-            // The first invariant is enforced by calling
-            // 'remote_timed_out_write_requests' in the same time slice as
-            // collecting the write requests. The second invariant is enforced
-            // by the strict order in which the ack() method is called
-            // explicitly after the operation is either committed or failed.
+        remove_timed_out_write_requests();
 
-            remove_timed_out_write_requests();
+        auto list = get_write_requests(
+          10_MiB); // TODO: use configuration parameter
 
-            auto list = get_write_requests(
-              10_MiB); // TODO: use configuration parameter
+        if (list.ready.empty()) {
+            co_return true;
+        }
 
-            if (list.ready.empty()) {
-                continue;
-            }
+        details::aggregator<Clock> aggregator;
+        while (!list.ready.empty()) {
+            auto& wr = list.ready.back();
+            wr._hook.unlink();
+            aggregator.add(wr);
+        }
+        // TODO: skip waiting if list.completed is not true
+        auto payload = aggregator.prepare();
+        auto result = co_await upload_object(
+          aggregator.get_object_id(), std::move(payload));
+        if (result.has_error()) {
+            // TODO: fix the error
+            // NOTE: it should be possible to translate the
+            // error to kafka error at the call site but I
+            // don't want to depend on kafka layer directly.
+            // Timeout should work well at this point.
+            aggregator.ack_error(errc::timeout);
+            co_return result.error();
+        }
+        aggregator.ack();
+        co_return list.complete;
+    } catch (...) {
+        auto err = std::current_exception();
+        if (ssx::is_shutdown_exception(err)) {
+            vlog(_logger.debug, "Batcher shutdown error: {}", err);
+            co_return errc::shutting_down;
+        }
+        vlog(_logger.error, "Unexpected batcher error: {}", err);
+        co_return errc::unexpected_failure;
+    }
+    __builtin_unreachable();
```

Comment:
is the compiler really complaining about this? probably worth removing it--pretty sure it disables real warnings

Reply:
the intention is to make this less surprising, when I'm seeing that there is no return at the end I have to ask myself is it by mistake or is it intentional? `__builtin_unreachable` removes this doubt